volatility3/framework/plugins/windows/mftscan.py

```diff
@@ -189,3 +173,143 @@ def run(self):
+                "ATTRIBUTE": mft.MFTAttribute,
+            },
+        )
+
+        # get each of the individual Field Sets
+        mft_object = symbol_table + constants.BANG + "MFT_ENTRY"
+        attribute_object = symbol_table + constants.BANG + "ATTRIBUTE"
+        fn_object = symbol_table + constants.BANG + "FILE_NAME_ENTRY"
+
+        # Scan the layer for Raw MFT records and parse the fields
+        for offset, _rule_name, _name, _value in layer.scan(
+            context=self.context, scanner=yarascan.YaraScanner(rules=rules)
+        ):
+            with contextlib.suppress(exceptions.PagedInvalidAddressException):
+                mft_record = self.context.object(
+                    mft_object, offset=offset, layer_name=layer.name
+                )
+                # We will update this on each pass in the next loop and use it as the new offset.
+                attr_base_offset = mft_record.FirstAttrOffset
+
+                attr = self.context.object(
+                    attribute_object,
+                    offset=offset + attr_base_offset,
+                    layer_name=layer.name,
+                )
+
+                # There is no field that has a count of Attributes
+                # Keep Attempting to read attributes until we get an invalid attr.AttrType
+                is_ads = False
+                file_name = renderers.NotAvailableValue
+                # The First $DATA Attr is the 'principal' file itself not the ADS
+                while attr.Attr_Header.AttrType.is_valid_choice:
+                    if attr.Attr_Header.AttrType.lookup() == "FILE_NAME":
+                        attr_data = attr.Attr_Data.cast(fn_object)
+                        file_name = attr_data.get_full_name()
+                    if attr.Attr_Header.AttrType.lookup() == "DATA":
+                        if is_ads:
+                            if not attr.Attr_Header.NonResidentFlag:
+                                # Resident files are the most interesting.
+                                if attr.Attr_Header.NameLength > 0:
+                                    ads_name = attr.get_resident_filename()
+                                    if not ads_name:
+                                        ads_name = renderers.NotAvailableValue
+
+                                    content = attr.get_resident_filecontent()
+                                    if content:
+                                        # Preparing for Disassembly
+                                        architecture = layer.metadata.get(
+                                            "architecture", None
+                                        )
+
+                                        disasm = (
+                                            interfaces.renderers.Disassembly(
+                                                content, 0, architecture.lower()
+                                            )
+                                            if architecture
```

Comment:
I think this would be easier to read as:

```
disasm = interfaces.renderers.BaseAbsentValue
if architecture:
   disasm = interfaces.renderers.Disassembly(content, 0, architecture.lower())
```

Reply:
Also, I get that our assumption is that ADS is most commonly used for hiding executable code, but could this not give a lot of false positives if the data stores in it is just data?  Also there seems to be no limit on the amount of data that might be disabled?  I assume files that are too large aren't stored in memory, but I just want to know the largest amount of data we might expect to find there, and how it might affect the output in volatility?

Reply:
I understand, 
So, the ADS is mainly used legitimately by Microsoft to introduce for example the Zone.Identifier feature, which is a nice digital forensics clear text artefact to identify the origin of a file. This will be for me our first assumption. Indeed, the value of the ZoneID indicate whether a file is from the internet (3), trusted(2), untrusted(4) etc.  
This is a nice pivot point which is the reason for the Hexdump view in order for the analyst to check it in clear text.

A derived usage of this feature can indeed be from an adversary trying to hide malicous code like you said (our second assumption). Although there is some false positives when an ADS doesn't contain any clear text data, the analyst might want to check that the name of the ADS is [a common data stream used by Windows](https://flatcap.github.io/linux-ntfs/ntfs/attributes/data.html) and the disam view to identify potential code. But it not a common practice to put code in the ADS. 

The largest amount of data you might expect per record is ~600 bytes or less for resident data (the remaining space left of the MFT entry which total size is 1024 bytes). This cannot be larger for a resident file. I think it will not affect the output in volatility3 😊